src/kernel/src/obj/copy.rs

```diff
@@ -24,86 +27,489 @@ fn split_range(
+                    let start_diff = last_point.offset(1) - last_range.start;
+                    let len_diff = last_range.length - start_diff;
+
+                    if last_range.length > len_diff {
+                        last_range.length -= len_diff;
+                        last_range.start = last_range.start.offset(start_diff);
+                        last_range.offset += start_diff;
+                        assert!(last_range.start == last_point.offset(1));
+                        last_range.gc_pagevec();
+                    
+                        let kicked = dest_tree.insert_replace(last_range.range(), last_range);
+                        assert!(kicked.is_empty());
+                    }
+                }
+        }
+
+        // Handle the first range, truncating it if it starts before the zeroing region. Don't bother removing it from
+        // the list -- we'll just skip it in the iterator (remove head of vec can be slow).
+        if let Some(first) = points.first() && 
+            let Some(mut first_range) = dest_tree.remove(first) {
+                let first_point = dest_point;
+                if first_point < first_range.start.offset(first_range.length) && first_point >= first_range.start {
+                    let len_diff = first_range.start.offset(first_range.length) - first_point;
+
+                    if first_range.length > len_diff {
+                        first_range.length -= len_diff;
+                        first_range.gc_pagevec();
+                        let kicked = dest_tree.insert_replace(first_range.range(), first_range);
+                        assert!(kicked.is_empty());
+                    }
+                }
+        }
+
+        // Finally we can remove the remaining ranges that are wholely contained. Skip the first one, though, we handled that above.
+        for point in points.iter().skip(1) {
+            dest_tree.remove(point);
+        }
+        dest_point = dest_point.offset(vec_pages);
+    }
+    remaining_pages -= vec_pages;
+
+    assert_eq!(remaining_pages == 1, end_offset > 0);
+    assert!(remaining_pages == 1 || remaining_pages == 0);
+
+    if end_offset > 0 {
+        zero_single(&mut dest_tree, dest_point, 0, end_offset);
+    }
 }
 
-impl CopySpec {
-    pub fn new(
-        src: ObjectRef,
-        src_start: PageNumber,
-        dest_start: PageNumber,
-        length: usize,
-    ) -> Self {
-        Self {
-            src,
-            src_start,
-            dest_start,
-            length,
+#[cfg(test)]
+mod test {
+    use twizzler_abi::{device::CacheType, object::Protections};
+
+    use crate::{
+        memory::context::{kernel_context, KernelMemoryContext, ObjectContextInfo},
+        obj::{pages::Page, ObjectRef, PageNumber, copy::zero_ranges},
+        userinit::create_blank_object,
+    };
+
+    use super::copy_ranges;
+
+    fn copy_ranges_and_check(
```

Comment:
Is that different from the test on line 510?

Reply:
I think it is. Each of these tests copies the data of a _single_ range and then checks. What I am asking is if we can copy multiple different ranges, and then check for equivalence, instead of one at a time. 

In practice, copying regions from ARM ELF files leads to copying from partial pages to other non-aligned offsets, and some regions are contiguously copied. Part of the reason the old stuff did not work is because whole pages were written, and sometimes old data copied over was overwritten.

Say we copy a range like so
```
p1_start = 0x50
p1_len = 0x270
p2_len = 0x4e8e4
p2_start = p1_start + p1_len
(Page 1 + p1_start, p1_len), (Page 1 + p1_start, p2_len) -> (Page 3, p1_len), (Page 3 + p2_start, p2_len) 
// check equivalence here
```

It doesn't have to be this test case (p2_len is really big), but something like this came up with ELF loading.

Reply:
I see, lemme take a stab at it.